### PR TITLE
Add deprecation warnings for pyopenssl

### PR DESCRIPTION
**Pull request by IvanLauLinTiong**
This PR is to address issue #2691

### CODE REVIEW

**Comment by sethmlarson** [comment]
@IvanLauLinTiong Looks like the "lint" job is failing, can you add a `# noqa: F401` comment in front of the import and run `nox -rs format` locally?

**Comment by IvanLauLinTiong** [comment]
> @IvanLauLinTiong Looks like the "lint" job is failing, can you add a `# noqa: F401` comment in front of the import and run `nox -rs format` locally?

Hi @sethmlarson,  It seems the `format' job fails to run when I running `nox -rs format` on my Window machine. I opened an issue here #2709 .

File: src/urllib3/__init__.py [hunk 1]
```diff
@@ -20,6 +20,19 @@
 from .util.retry import Retry
 from .util.timeout import Timeout
 
+try:
+    import urllib3_secure_extra  # type: ignore # noqa: F401
+except ImportError:
```
Comment (graingert) [inline src/urllib3/__init__.py:25]: I think in 3.7+?

```suggestion
except ModuleNotFoundError:
```

File: src/urllib3/__init__.py [hunk 2]
```diff
@@ -20,6 +20,19 @@
 from .util.retry import Retry
 from .util.timeout import Timeout
 
+try:
```
Comment (graingert) [inline src/urllib3/__init__.py:23]:
```suggestion
# note to vendors and OS distributors delete this block
try: 
```

File: src/urllib3/contrib/pyopenssl.py [hunk 3]
```diff
@@ -53,13 +53,22 @@ class UnsupportedExtension(Exception):  # type: ignore[no-redef]
...
 import ssl
+import warnings
 from io import BytesIO
 from socket import socket as socket_cls
 from socket import timeout
 from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Union
 
 from .. import util
 
+warnings.warn(
```
Comment (graingert) [inline src/urllib3/contrib/pyopenssl.py:64]: You might want this warning only when you construct PyOpenSSLContext or call `inject_into...`

File: src/urllib3/__init__.py (hunk 1 above)
Comment (IvanLauLinTiong) [reply src/urllib3/__init__.py:25]: According to Python [doc](https://docs.python.org/3/library/exceptions.html#ModuleNotFoundError), `ModuleNotFoundError` is a subclass of `ImportError`.  Ya I agree with you,  `ModuleNotFoundError` seems more intuitive and serves right purpose here though `except ImportError` still able to catch the exception.

Comment (graingert) [reply src/urllib3/__init__.py:25]: The purpose of ModuleNotFoundError is to distinguish cases where something breaks in a significant way from the module simply not existing.

If an ImportError that's not a ModuleNotFoundError shows up here then users will want to know about it

File: src/urllib3/__init__.py (hunk 2 above)
Comment (IvanLauLinTiong) [reply src/urllib3/__init__.py:24]: @graingert  I don't understand your question. What does it mean `we don't want pip itself to hit this`? 
What this block of code doing here is to try importing the marker package `urllib3-secure-extra` (which indicate `secure` extra has installed), if import success, we raise the `DeprecationWarning` message.

Comment (graingert) [reply src/urllib3/__init__.py:24]: so urllib3 is pasted wholesale into `pip._vendor` and we don't want installing a package in someone's virtual environment to start causing pip issuing warnings

Additionally some distributors like RedHat/fedora/canonical/debian will redistribute urllib3 and they might not want it issuing warnings to their users. We certainly don't want them to attempt to distribute the marker package.

Doubtless they'll make the right choice here even if we don't add a note here

Comment (sethmlarson) [reply src/urllib3/__init__.py:27]: Excellent idea, thanks @graingert!

**Review by sethmlarson — DISMISSED** [review]
LGTM, now let's backport this to 1.26.x branch. We'll have to use `ImportError` there to support Python versions earlier than 3.7.

**Comment by graingert** [comment]
```
if sys.version_info < (3, 6):
    ModuleNotFoundError = ImportError
```

**Comment by IvanLauLinTiong** [comment]
> [benjaminp/six#201](https://github.com/benjaminp/six/issues/201)
> 
> ```
> if sys.version_info < (3, 6):
>     ModuleNotFoundError = ImportError
> ```

Not bad! we can add this checking

**Review by sethmlarson — APPROVED** [review]

**Comment by sethmlarson** [comment]
There was a coverage issue when we switched ModuleNotFoundError -> ImportError but I've fixed it on your branch. Going to merge once we get a green light :)

**Review by sethmlarson — APPROVED** [review]

**Comment by sethmlarson** [comment]
Thanks @IvanLauLinTiong, can you backport this PR to 1.26.x and then you can claim your bounty :)

**Comment by IvanLauLinTiong** [comment]
> Thanks @IvanLauLinTiong, can you backport this PR to 1.26.x and then you can claim your bounty :)

Hi @sethmlarson. I already created the  [PR](https://github.com/urllib3/urllib3/pull/2701) to backport the changes to 1.26.x.

**Pull request merged**